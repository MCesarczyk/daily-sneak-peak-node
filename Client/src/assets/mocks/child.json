{
  "id": 2,
  "name": "Mary",
  "surname": "Jane",
  "group": "Crocodiles",
  "created_at": "2022-01-24T00:08:31.474Z",
  "updated_at": "2022-01-24T00:08:31.474Z"
}