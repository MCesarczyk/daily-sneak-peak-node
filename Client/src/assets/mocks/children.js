export const children = [
  {
      "id": 6,
      "name": "Amy",
      "surname": "Raven",
      "group": "Monkeys",
      "created_at": "2022-01-27T08:55:07.942Z",
      "updated_at": "2022-01-27T08:55:07.942Z"
  },
  {
      "id": 1,
      "name": "Jack",
      "surname": "Smith",
      "group": "Frogs",
      "created_at": "2022-01-24T00:08:16.967Z",
      "updated_at": "2022-01-24T00:08:16.967Z"
  },
  {
      "id": 3,
      "name": "John",
      "surname": "Doe",
      "group": "Crocodiles",
      "created_at": "2022-01-24T00:08:55.735Z",
      "updated_at": "2022-01-24T00:09:08.131Z"
  },
  {
      "id": 2,
      "name": "Mary",
      "surname": "Jane",
      "group": "Crocodiles",
      "created_at": "2022-01-24T00:08:31.474Z",
      "updated_at": "2022-01-24T00:08:31.474Z"
  },
  {
      "id": 4,
      "name": "Peppa",
      "surname": "Pig",
      "group": "Frogs",
      "created_at": "2022-01-24T22:37:07.096Z",
      "updated_at": "2022-01-24T22:37:07.096Z"
  },
  {
      "id": 5,
      "name": "Sophie",
      "surname": "Moon",
      "group": "Monkeys",
      "created_at": "2022-01-27T08:52:45.470Z",
      "updated_at": "2022-01-27T08:52:45.470Z"
  }
]