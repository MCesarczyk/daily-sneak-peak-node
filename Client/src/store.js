import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "@redux-saga/core";
import childrenReducer from "./views/children/childrenSlice";
import rootSaga from "./rootSaga";

const sagaMiddleware = createSagaMiddleware();

const store = configureStore({
  reducer: {
    children: childrenReducer,
  },
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);

export default store;