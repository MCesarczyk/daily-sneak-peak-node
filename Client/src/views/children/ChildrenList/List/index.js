import List from '@mui/material/List';
import ListViewItem from './Item';
import { Space } from '../../../../components/Space';
import Loader from '../../../../components/Loader';

const ListView = ({ state, childrenList }) => (
  <Loader
    state={state}
    message="Loading"
  >
    <Space>
      <List sx={{ width: '100%', bgcolor: 'background.paper' }}>
        {childrenList.length > 0 && childrenList.map(child => (
          <ListViewItem
            key={child.id}
            child={child}
          />
        ))}
      </List>
      <Space
        vertical
        justify="start"
      >
      </Space>
    </Space>
  </Loader>
);

export default ListView;