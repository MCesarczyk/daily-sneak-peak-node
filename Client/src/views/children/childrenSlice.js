import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  children: [],
  state: 'idle'
};

const childrenSlice = createSlice({
  name: 'children',
  initialState,
  reducers: {
    fetchChildrenList: () => { },
    setChildrenList: (state, { payload: children }) => {
      state.children = children;
    },
    reloadChildrenList: () => initialState,
    clearChildrenList: () => initialState,
    setChildrenState: (state, {payload: newState}) => {
      state.state = newState;
    }
  },
});

export const {
  fetchChildrenList,
  setChildrenList,
  reloadChildrenList,
  clearChildrenList,
  setChildrenState
} = childrenSlice.actions;

const selectChildren = state => state.children;
export const selectChildrenList = state => selectChildren(state).children;
export const selectChildrenState = state => selectChildren(state).state;

export default childrenSlice.reducer;