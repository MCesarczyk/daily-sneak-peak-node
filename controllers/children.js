const { pool } = require('../db');

const getAllChildren = (req, res, next) => {
  pool.query(`SELECT * FROM children
  ORDER BY created_at DESC`,
    (q_err, q_res) => {
      res.json(q_res.rows)
    })
};

const createChild = (req, res, next) => {
  const values = [req.body.name,
  req.body.surname,
  req.body.group,
  req.body.avatar]

  pool.query('INSERT INTO children (name, surname, "group", created_at, updated_at, avatar)\
   VALUES ($1, $2, $3, NOW(), NOW(), $4)',
    values, (q_err, q_res) => {
      if (q_err) return next(q_err);
      res.json(q_res.rows)
    })
};

const getOneChild = (req, res, next) => {
  const child_id = req.params.id

  pool.query(`SELECT * FROM children
              WHERE id=$1`,
    [child_id], (q_err, q_res) => {
      res.json(q_res.rows)
    })
};

const updateChild = (req, res, next) => {
  const values = [req.body.name,
  req.body.surname,
  req.body.group,
  req.params.id]

  pool.query(`UPDATE children SET name=$1, surname=$2, "group"=$3, updated_at=NOW()\
   WHERE id=$4`,
    values, (q_err, q_res) => {
      if (q_err) return next(q_err);
      res.json(q_res.rows)
    })
};

const deleteChild = (req, res, next) => {
  const child_id = req.params.id

  pool.query(`DELETE FROM children WHERE id = $1`,
    [child_id], (q_err, q_res) => {
      res.json(q_res.rows)
      console.log(q_err)
    })
};

module.exports = { getAllChildren, createChild, getOneChild, updateChild, deleteChild };