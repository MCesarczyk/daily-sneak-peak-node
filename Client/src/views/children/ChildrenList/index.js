import { useEffect } from 'react';
import { useSelector } from "react-redux";
import { selectChildrenList, selectChildrenState } from "../childrenSlice";
import { useDispatch } from 'react-redux';
import { clearChildrenList, fetchChildrenList } from '../childrenSlice';
import ListView from './List';

const ChildrenList = () => {
  const childrenList = useSelector(selectChildrenList);
  const state = useSelector(selectChildrenState);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchChildrenList());

    return (() => {
      dispatch(clearChildrenList());
    });
  }, []);

  return (
    <ListView
      state={state}
      childrenList={childrenList}
    />
  )
};

export default ChildrenList;