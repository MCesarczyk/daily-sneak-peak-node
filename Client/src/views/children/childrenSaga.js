import { call, put, delay, takeLatest } from "redux-saga/effects";
import { DEMO_DELAY } from "../../assets/data";
import { CHILDREN_LIST_URL } from "../../assets/links";
import { getDataFromApi } from "../../assets/utils/handleApiCalls";
import { fetchChildrenList, reloadChildrenList, setChildrenList, setChildrenState } from "./childrenSlice";

function* fetchChildrenListHandler() {
  try {
    const url = CHILDREN_LIST_URL;
    yield put(setChildrenState("loading"));
    const response = yield call(getDataFromApi, url);
    const data = yield response;
    yield delay(DEMO_DELAY);
    yield put(setChildrenList(data));
    yield put(setChildrenState("success"));
  } catch (error) {
    yield call(console.error, error.message);
  }
};

export function* childrenSaga() {
  yield takeLatest(fetchChildrenList.type, fetchChildrenListHandler);
  yield takeLatest(reloadChildrenList.type, fetchChildrenListHandler);
};