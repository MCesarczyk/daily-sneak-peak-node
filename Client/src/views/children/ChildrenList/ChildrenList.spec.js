import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import store from "../../../store";
import { children } from "../../../assets/mocks/children";
import reducer from "../childrenSlice";
import ChildrenList from ".";

test("Children list renders correctly", () => {
  render(
    <Provider store={store}>
      <ChildrenList />
    </Provider>
  );

  const list = screen.getByTestId("childrenList");
  expect(list).toBeInTheDocument();
});

test("Children reducer should return initial state", () => {
  expect(reducer(undefined, {})).toEqual({
    "children": []
  });
});

test("Children reducer should return stored data", () => {
  const previousState = children;
  expect(reducer(previousState, {})).toMatchSnapshot();
});