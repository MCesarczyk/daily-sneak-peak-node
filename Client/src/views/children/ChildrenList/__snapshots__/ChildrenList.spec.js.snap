// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Children reducer should return stored data 1`] = `
Array [
  Object {
    "created_at": "2022-01-27T08:55:07.942Z",
    "group": "Oposy",
    "id": 6,
    "name": "Ala",
    "surname": "Rawecka",
    "updated_at": "2022-01-27T08:55:07.942Z",
  },
  Object {
    "created_at": "2022-01-24T00:08:16.967Z",
    "group": "Żabki",
    "id": 1,
    "name": "Jack",
    "surname": "Smith",
    "updated_at": "2022-01-24T00:08:16.967Z",
  },
  Object {
    "created_at": "2022-01-24T00:08:55.735Z",
    "group": "Krokodylki",
    "id": 3,
    "name": "John",
    "surname": "Doe",
    "updated_at": "2022-01-24T00:09:08.131Z",
  },
  Object {
    "created_at": "2022-01-24T00:08:31.474Z",
    "group": "Krokodylki",
    "id": 2,
    "name": "Mary",
    "surname": "Jane",
    "updated_at": "2022-01-24T00:08:31.474Z",
  },
  Object {
    "created_at": "2022-01-24T22:37:07.096Z",
    "group": "Żabki",
    "id": 4,
    "name": "Peppa",
    "surname": "Pig",
    "updated_at": "2022-01-24T22:37:07.096Z",
  },
  Object {
    "created_at": "2022-01-27T08:52:45.470Z",
    "group": "Oposy",
    "id": 5,
    "name": "Zosia",
    "surname": "Sebastiańska",
    "updated_at": "2022-01-27T08:52:45.470Z",
  },
]
`;
